M2/Macaulay2/packages/SuperLinearAlgebra.m2

```diff
@@ -0,0 +1,736 @@
+newPackage(
+    "SuperLinearAlgebra", 
+    Version => "0.1", 
+    Date => "29 January 2021", 
+    Authors => {
+        {Name => "Fereshteh Bahadorykhalily", 
+         Email => "f.bahadori.khalili@gmail.com", 
+         HomePage => "https://www.researchgate.net/profile/Fereshte_Bahadorykhalily"
+        }, 
+        {Name => "Fatemeh Tarashi Kashani", 
+         Email => "tarashikashanifatemeh@gmail.com", 
+         HomePage => "https://www.linkedin.com/in/fatemehtarashi/"
+        }
+    }, 
+    Headline => "berezinian and supertrace of a supermatrix", 
+    DebuggingMode => false
+)
+
+--------------------
+-- Exports
+--------------------
+
+export {
+   -- Types
+    "SuperMatrix", 
+    
+   -- Methods
+    "berezinian", 
+    "inverseSuperMatrix", 
+    "parity", 
+    "superMatrixGenerator", 
+    "superMatrixParity", 
+    "superRing", 
+    "superTrace", 
+    
+   -- Symbols
+    "supermatrix", 
+    "supermatrixFirstBlockTarget", 
+    "supermatrixSecondBlockTarget", 
+    "supermatrixFirstBlockSource", 
+    "supermatrixSecondBlockSource"
+}
```

Comment:
What do "First" and "Second" mean here?  (Maybe I can help make a better name.)

Reply:
Each supermatrix is made with 4 matrices(4 blocks).
supermatrix =  
| M1 | M2 |
|  --- | ---  |
| M3 | M4 |

The key supermatrixFirstBlockTarget shows the number of first part rows. 
The key supermatrixFirstBlockSource shows the number of columns in the first part.  
Similarly named for the second block.

Reply:
That numbering seems sort of clumsy to me.
Maybe this would make more sense:

| 0,0 | 0,1 |
|  --- | ---  |
| 1,0 | 1,1 |

Reply:
Thanks. I'll change the names in the function and docs. ‌But for these keys, What do you think is right now to name them sourceM00 and targetM01?

Reply:
No, those names are not good.  What about using notation?  If the supermatrices belong to a new type that you've created, then you could install methods so that `M_(i,j)` could give the correspnding block.  But if you want that to actually give the entry in row i column j instead of the block, maybe mimic what M2 already does for block matrices:
```m2
Macaulay2, version 1.17
with packages: ConwayPolynomials, Elimination, IntegralClosure, InverseSystems, LLLBases,
               MinimalPrimes, PrimaryDecomposition, ReesAlgebra, Saturation, TangentCone

i1 : f = matrix {{1,2},{3,4}}

o1 = | 1 2 |
     | 3 4 |

              2        2
o1 : Matrix ZZ  <--- ZZ

i2 : g = matrix {{f,2*f},{3*f,4*f}}

o2 = | 1 2  2  4  |
     | 3 4  6  8  |
     | 3 6  4  8  |
     | 9 12 12 16 |

              4        4
o2 : Matrix ZZ  <--- ZZ

i5 : g_[1]^[1]

o5 = | 4  8  |
     | 12 16 |

              2        2
o5 : Matrix ZZ  <--- ZZ

i6 : g_[0]^[1]

o6 = | 3 6  |
     | 9 12 |

              2        2
o6 : Matrix ZZ  <--- ZZ
```

Reply:
Thanks. We hadn't seen this before. I think this is much better. I'll rewrite the functions with this. 
See here @fereshteh-bahadory.

Reply:
Dear @DanGrayson,
Thanks again for your comments and consideration.
We tried to replace the superMatrix with the method that you kindly suggested. This works for superMatrix itself, but when we want to use it in another function, superMatrixParity, we get an error "stdio:75:44:(3): error: expected first parameter to be a hash table". I am not sure, but I think it is better if we use the method which you wrote in our hash-table. The following may work:

SuperMatrix = new Type of MutableHashTable;
superMatrixGenerator = method();
superMatrixGenerator (Matrix, Matrix, Matrix, Matrix) := (M00,M01,M10,M11)-> (
    nr1 := numgens target M00;
    nr2 := numgens target M10;
    ns1 := numgens source M00;
    ns2 := numgens source M11;
    new SuperMatrix from {
        superMatrix => matrix{{M00,M01},{M10,M11}}, 
        target00 => nr1, 
        target10 => nr2, 
        source00 => ns1, 
        source10 => ns2
    }
)

Reply:
If you can't figure out where the error is coming from, push the offending code to another branch and I'll take a look at it for you.

Reply:
Thank you very much. The problem has been solved. We have updated the file regarding your comments.